modules/sitemaps/sitemaps.php

```diff
@@ -1,782 +1,1556 @@
+			$buffer->try_to_add_item(
+				"<sitemap>\n" .
+ 				" <loc>" . home_url() . "/sitemap-index-$i.xml</loc>\n" .
+				" <lastmod>" . str_replace( ' ', 'T', $timestamp) . 'Z' . "</lastmod>\n" .
+				"</sitemap>\n"
+			);
+		}
 
-	$post_types_in = array();
-	foreach ( (array) $post_types as $post_type ) {
-		$post_types_in[] = $wpdb->prepare( '%s', $post_type );
-	}
-	$post_types_in = join( ",", $post_types_in );
+		// Until the buffer is too large,
+		while ( false == $buffer->is_full() ) {
+			// Retrieve a batch of posts (in order)
+			$posts = $this->query_page_sitemaps_after_ID($last_sitemap_ID, 1000);
 
-	// use direct query instead because get_posts was acting too heavy for our needs
-	//$posts = get_posts( array( 'numberposts'=>1000, 'post_type'=>$post_types, 'post_status'=>'published' ) );
-	$posts = $wpdb->get_results( "SELECT ID, post_type, post_modified_gmt, comment_count FROM $wpdb->posts WHERE post_status='publish' AND post_type IN ({$post_types_in}) ORDER BY post_modified_gmt DESC LIMIT 1000" );
-	if ( empty( $posts ) ) {
-		status_header( 404 );
-	}
-	header( 'Content-Type: ' . jetpack_sitemap_content_type() );
-	$initstr = jetpack_sitemap_initstr( get_bloginfo( 'charset' ) );
-	$tree    = simplexml_load_string( $initstr );
-	// If we did not get a valid string, force UTF-8 and try again.
-	if ( false === $tree ) {
-		$initstr = jetpack_sitemap_initstr( 'UTF-8' );
-		$tree    = simplexml_load_string( $initstr );
-	}
+			// If there were no posts to get, make a note.
+			if (null == $posts) {
+				$any_sitemaps_left = false;
+				break;
+			}
 
-	unset( $initstr );
-	$latest_mod = '';
-	foreach ( $posts as $post ) {
-		setup_postdata( $post );
+			// Otherwise, for each post in the batch,
+			foreach ($posts as $post) {
+				// Generate the sitemap XML for the post.
+				$current_item = $this->sitemap_to_index_item($post);
+
+				// If adding this item to the buffer doesn't make it too large,
+				if ( true == $buffer->try_to_add_item($current_item['xml']) ) {
+					// Add it and update the last sitemap ID.
+					$last_sitemap_ID = $post->ID;
+					$buffer->view_time($current_item['last_modified']);
+				} else {
+					// Otherwise stop looping through posts.
+					break;
+				}
+			}
+		}
 
-		/**
-		 * Filter condition to allow skipping specific posts in sitemap.
-		 *
-		 * @module sitemaps
-		 *
-		 * @since 3.9.0
-		 *
-		 * @param bool $skip Current boolean. False by default, so no post is skipped.
-		 * @param WP_POST $post Current post object.
+		// Store the buffer as the content of a jp_sitemap_index post.
+		$this->set_contents_of_post(
+			'sitemap-index-' . $number,
+			'jp_sitemap_index',
+			$buffer->contents(),
+			$buffer->last_modified()
+		);
+
+		/*
+		 * Now report back with the ID of the last sitemap post ID to
+		 * be successfully added, whether there are any sitemap posts
+		 * left, and the most recent modification time seen.
 		 */
-		if ( apply_filters( 'jetpack_sitemap_skip_post', false, $post ) ) {
```

Comment:
It'd be nice to keep that filter around somehow, it's useful for some site owners.